### DISCUSSION TITLE
Array Variable Solution IC

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#26932] [Q&A General] [answered]
Greetings everyone,

I wanted to ask about the possibility of using the SolutionIC / SolutionUserObject procedure to initialize an array variable.
It seems to me that, at the moment, this is not possible ( "// We can handle only FIRST order scalar variables" is written in the SolutionUserObject source file, and if I try to follow the example at /moose/test/tests/ics/solution_ic/solution_ic.i with array variable, I get a "_Map_base::at" error).

Is there any other procedure that is standard for this purpose?
I realize that one could for instance use an "ArrayConstantIC", but I would much rather be able to just carry over solution values from a previous simulation.
For context, I am looking to use the solution of the k eigenproblem for the multigroup diffusion equation as the initial condition for a transient, where the initial precursor distribution is calculated with an auxkernel.

In case the solution would indeed be to produce an array version of SolutionIC / SolutionUserObject, I could try to take a stab at it. Is there any reason this would be impractical?

Thanks for your time,

Gabriele

**Comment by GiudGiud** [comment, marked as answer]
Hello

You probably wont want to initialize every variable manually from SolutionICs + a SolutionUO, even though I think we should be able to make it work with a (newly created) SolutionArrayIC.

I would recommend trying either of these two techniques to initialize everything at once:

https://mooseframework.inl.gov/application_usage/restart_recover.html

Guillaume

**Reply by GiudGiud** [reply]
if you want to code up the SolutionArrayIC we would love to have it in moose btw

**Reply by gabburgio** [reply]
Thanks for the solution.
I do have a followup question, since it appears that now the flux variables are being correctly copied from the checkpoint file, but the precursor auxiliary variables are not.

By using postprocessors to compute the integrals, I can tell that the precursor concentrations are all set to zero in the beginning of the transient simulation (unlike the fluxes, whose integral are the same as those in TIMESTEP_END of the eigenvalue calculation).
Is this due to the fact that I am trying to copy auxiliary variables into normal ones?
Is there any other syntax/flag that I should add in this case?

I am using what I think are the standard advanced restart commands: 
```
[Outputs]
[cp]
type = Checkpoint
execute_on = TIMESTEP_END 
[]
[]
```
```

[Mesh]
file = eigen_cp_cp/0001-mesh.cpr
[]

[Problem]
restart_file_base = eigen_cp_cp/LATEST  
[]
```

Also this doesn't seem to be due to the execution flag on the auxiliary kernels for the precusors in the eigenvalue calculation, which I have tried changing without result.
I would be very grateful for any indication about what's going wrong.
Thanks again,

Gabriele

**Reply by GiudGiud** [reply]
Hello

Can you please paste the auxkernel and auxvariables block?
The change of system is indeed the problem.
I think the ArraySolutionIC is the solution for this

Guillaume

**Reply by YaqiWang** [reply]
>Is this due to the fact that I am trying to copy auxiliary variables into normal ones?

Are you saying precursor concentrations are aux variables in the eigenvalue calculation but primal variables during transient? If so, I think the restart system cannot handle this. @GiudGiud has commented. I think @loganharbour will be working on a capability that allows users to restart individual items from a checkpoint file, which will have a similar syntax like `Variables/*/initial_from_file_var`. But you may need to wait for a while on this.

**Reply by gabburgio** [reply]
Thank you both for the information, indeed it is the case that precursor would have gone from auxiliary to primal
